Constructions/QuotationTactics.ml

```diff
@@ -1,29 +1,188 @@
+needs "Constructions/epsilon.ml";;
+
 (*To make proofs easier, this function attempts to automatically match up string equalities, to do so,
 it generates a list of calls to STRING_EQ_CONV for all strings in the term which can then be used with
 REWRITE_TAC*)
 
-let rec STRING_FETCH tm = match tm with
-	| Comb(Comb(Const("=",_),a),b) when type_of a = mk_type("list",[mk_type ("char",[])]) -> [STRING_EQ_CONV(tm)]
-	| Comb(a,b) -> (STRING_FETCH a) @ (STRING_FETCH b)
-	| Abs(a,b) -> (STRING_FETCH a) @ (STRING_FETCH b)
-	| Quote(e) -> (STRING_FETCH e)
-	| Hole(e,ty) -> (STRING_FETCH e)
-	| Eval(e,ty) -> (STRING_FETCH e)
-	| _ -> [];; 
+let rec STRING_FETCH tm = 
```

Comment:
I'm concerned about the need for this. Why are 'strings' being manipulated like this at all? [There might be good reasons, but just as often, it's a sign of hacking]

Reply:
Internally, you have to have some code manipulate low-level representations. But only internal routines should ever call such low-level manipulations. This shouldn't leak out to the user. [ And if STRING_EQ_CONV is user-visible, I would consider that to be a bug on Harrison's part]

If it's just about T and F, maybe it's better to hard-code that, than to let some general name-revealing routines leak out too far.

Reply:
Ok I will hard-code it in.

Reply:
This is going to take longer than I thought- lots of the functions are defined in terms of comparing strings. 
Most of the changes will result in no longer being able to prove formulas false- which I guess is fine, I'll just remove the tests that evaluate to F.
I am unsure how to deal with the isFreeIn theorem though, as it requires checking that the variable is not the binding variable in an abstraction, which I can no longer do if I don't compare strings. 

Am I allowed to change Harrison's original tactics to automatically equate strings (for example REWRITE_TAC)? 
I feel like that's cheating though...
I'm not sure how to get around that issue.

Reply:
Defining some functions in terms of comparing strings is fine! What is not good is letting the implementations details leak.  So as long as `STRING_FETCH` is not user-accessible, that's fine.

Reply:
But if I made STRING_FETCH an internal function in another function (like REWRITE_TAC for ex.) isn't it still kind of "user-accessible"? Or is that not the right path.

I'm confused about how to do this.
(also, is the plan to never use STRING_EQ_CONV in a proof?)

Reply:
Values of type "string" are components of both types and terms.  So the proof developer needs to be able to access them. In this case, "string" is an external representation.

Reply:
I think that is too simplistic a view. It's like saying that organic molecules (eventually) contain quarks, and so these need to be accessible to do anything.

The proof developer should never need to see the implementation type "string" -- if they do, I would consider that a very severe bug. They may need to know that certain things have 'names' and that 'names' have an API. But that API should be very poor (a tiny fraction of the things you can actually do with strings) and abstract, in that string operations cannot be applied to names.

The whole point of the ML-HOL approach to theorem proving is that 'thm' is an abstract type.  While 'type' and 'term' are not quite as abstract, still, they do not expose their guts for all to see and play with.